homeassistant/components/scheduler/__init__.py

```diff
@@ -0,0 +1,506 @@
+            if (not schedule.end_datetime and schedule.start_datetime < now)
+            or (schedule.end_datetime and schedule.end_datetime < now)
+        }
+
+        await self.async_save()
+
+    async def async_create(
+        self,
+        entity_id: str,
+        start_datetime: datetime,
+        *,
+        end_datetime: Optional[datetime] = None,
+        rrule_str: Optional[str] = None,
+        schedule_id: Optional[str] = None,
+    ) -> dict:
+        """Create a schedule.
+
+        The rrule_str parameter must be an RFC5545-compliant string.
+        """
+        if not schedule_id:
+            schedule_id = uuid4().hex
+
+        schedule = Schedule(
+            self._hass,
+            schedule_id,
+            entity_id,
+            start_datetime,
+            end_datetime=end_datetime,
+            rrule_str=rrule_str,
+        )
+        self.schedules[schedule_id] = schedule
+
+        await self.async_save()
+        _LOGGER.info("Created schedule: %s", schedule)
+        await schedule.async_schedule_next_instance()
+
+        return schedule.as_dict()
+
+    async def async_delete(self, schedule_id: str, *, save: bool = True) -> dict:
+        """Delete a schedule."""
+        if schedule_id not in self.schedules:
+            raise KeyError
+
+        schedule = self.schedules.pop(schedule_id)
+        await schedule.active_instance.async_cancel()
+
+        if save:
+            await self.async_save()
+        _LOGGER.info('Deleted schedule "%s"', schedule)
+
+        return schedule.as_dict()
+
+    async def async_load(self) -> None:
+        """Load all schedules from storage."""
+        raw_schedules = await self._store.async_load()
+
+        if not raw_schedules:
+            return
+
+        tasks = []
+        for schedule_id, schedule in raw_schedules.items():
+            if schedule.get(CONF_END_DATETIME):
+                end_dt = dt_util.parse_datetime(schedule[CONF_END_DATETIME])
+            else:
+                end_dt = None
+            tasks.append(
+                self.async_create(
+                    schedule[CONF_ENTITY_ID],
+                    dt_util.parse_datetime(schedule[CONF_START_DATETIME]),
+                    end_datetime=end_dt,
+                    rrule_str=schedule.get(CONF_RRULE),
+                    schedule_id=schedule_id,
+                )
+            )
+
+        await asyncio.gather(*tasks)
+
+    async def async_save(self) -> None:
+        """Save all schedules to storage."""
+        await self._store.async_save(
+            {
+                schedule_id: schedule.as_dict()
+                for schedule_id, schedule in self.schedules.items()
+            }
+        )
+
+    async def async_update(self, schedule_id: str, new_data: dict) -> dict:
+        """Update a schedule."""
+        if schedule_id not in self.schedules:
+            raise KeyError
+
+        new_data = SCHEDULE_UPDATE_SCHEMA(new_data)
```

Comment:
This should not be necessary. Only websocket API cannot trust the data, after that, we trust that we get the right data.